Implement "exhaustive" option

Bellisario opened the pull request:
Closes #9.

P.S.: I focused on coding instead of adding an example on readme or new example on types... if I need provide some code examples, let me know!

sindresorhus commented on test.js:
```diff
@@ -23,3 +23,26 @@ test('main', t => {
 	t.pass();
 });
+
+test('preventRepeat', t => {
```
```suggestion
test('`preventRepeat` option', t => {
```

sindresorhus commented on index.js:
```diff
@@ -1,11 +1,25 @@
-export default function uniqueRandom(minimum, maximum) {
+function getMaxNumbers(minimum, maximum) {
```
Use a single-line arrow function.

sindresorhus commented on index.d.ts:
````diff
@@ -13,4 +13,4 @@ console.log(random(), random(), random());
 ```
 */
-export default function uniqueRandom(minimum: number, maximum: number): () => number;
+export default function uniqueRandom(minimum: number, maximum: number, preventRepeat?: boolean): () => number;
````
The option should be an options-object.

sindresorhus commented:
You will need to document it in the readme too.

Use this style: https://github.com/sindresorhus/boxen#options

sindresorhus commented:
I think the option needs a better name. It does actually allow repetition, just not until all the numbers have been used up.

sindresorhus commented on test.js:
```diff
@@ -23,3 +23,26 @@ test('main', t => {
+	}
+
+	t.pass();
+});
```
You need a test that actually ensures it does what it does. From what I can tell, this is just a copy of the previous test.

Bellisario commented:
I just updated with requested changes, let me know if there is something to correct.
I also changed the name from preventRepeat to option noOverlap, tell me if this is a good way to call it...
Now I'll focus on README.md updates... :smiley:

Richienb commented:
The original code called `random()` if the previous value matched the current one.

Where $n$ is the range of possible numbers, we can calculate how many times on average `random` is actually being called each time the consumer calls it, excluding the first call which doesn't have a previous value:

$$
c = \frac{1}{\frac{n - 1}{n}}
$$

where $\frac{n - 1}{n}$ calculates the chance of generating the number first try and we find the reciprocal to discover the amount of times we take that chance for it to theoretically be 100%.

For a range of 5 numbers, $c = 1.25$. This means each time `random` is called, it is actually called 1.25 times because of the recursion.

This amount decreases with the amount of numbers since there is a lower chance of collision. This means for $n = 50$, $c \approx 1.02$

The new algorithm has the inverse effect since this time, there are multiple previous numbers, increasing the chance of collision. To iterate over a set of 5 numbers, 5 calls to `random()` is optimal. However, because of the recursion, we can calculate the actual amount with:

$$
t = \sum_{i=1}^{n} \frac{1}{\frac{i}{n}}
$$

where $n$ is the range of numbers we are generating. For $n = 5$, $t \approx 11.42$ and for $n = 50$, $t \approx 224.96$.

Here, we graph $x$ as the range of numbers and $y$ as the amount of calls required:

<a href="https://www.desmos.com/calculator/i0djrbr3du"><img alt="Line graph containing the aforementioned figures" src="https://user-images.githubusercontent.com/29491356/180359935-85e76358-9197-4a26-a47f-df463022b7c9.svg" width="512" /></a>

As you can see, **the recursion is catastrophic**. The tests only check for a range of 1000 numbers, where $t \approx 7485.47$ which isn't large enough for the lag to be easily noticable so it hasn't been discovered experimentally.

I believe the most straightforward solution would be the Tetris sack system, whereby the numbers are generated beforehand with a shuffle and then pulled out of the "sack" one by one until empty.

Bellisario commented:
@Richienb, @sindresorhus, I updated the algorithm using the sack system to improve performance.
Now should be about comparable to the original random extraction, with the only limit of the initial Set configuration (for-loop) with all values can be extracted.

Richienb commented on index.js:
```diff
@@ -1,11 +1,37 @@
+		}
+
+		if (options?.noOverlap) {
+			const value = [...toExtract][Math.floor(Math.random() * toExtract.size)];
```
Just keep it as a regular array so that you don't need to be converting it every call.

Richienb commented on index.js:
```diff
@@ -1,11 +1,37 @@
+}
+
+export default function uniqueRandom(minimum, maximum, options) {
+	const defaultArray = filledArray(minimum, maximum);
```
Logic for `noOverlap` and the default behaviour can be seperated with a high-level if statement because of how different their implementations are.

sindresorhus commented on index.js:
```diff
@@ -1,4 +1,34 @@
-export default function uniqueRandom(minimum, maximum) {
+function filledArray(minimum, maximum) {
+	const array = [];
+	for (let i = minimum; i < maximum + 1; i++) {
```
`i` => `index`

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,7 @@
+interface Options {
+	noOverlap?: boolean;
```
You need to document this option.

See: https://github.com/sindresorhus/typescript-definition-style-guide

sindresorhus commented:
I don't think `noOverlap` is a good enough option name.

Bellisario commented:
> I don't think `noOverlap` is a good enough option name.

I don't know how we could name this option... :tired_face: seems very difficult to me, because it's particular...

Richienb commented:
`exhaustRange`?

`exhaustiveSequence`?

Bellisario commented:
Maybe also just `sack`, because of the sack system. (simpler, single word, and not a negation)...

sindresorhus commented on readme.md:
```diff
@@ -23,10 +23,32 @@ console.log(random(), random(), random());
+
+Type: `boolean`\
+Default: `false`\
+Prevent the same number from being returned twice in a row.
```
This is not what it does.

sindresorhus commented:
> Maybe also just sack, because of the sack system. (simpler, single word, and not a negation)...

I don't think most users of this package would know what "sack" is.

sindresorhus commented:
> I don't know how we could name this option...

A big part of submitting a pull request is docs and naming. Try to do some research ;)

Bellisario commented:
I found the technical name of this "sack system": `Shuffle Bag`.
Here is for example a C# article talking about it: https://gamedevelopment.tutsplus.com/tutorials/shuffle-bags-making-random-feel-more-random--gamedev-1249

The problem is that also this name is not so simple to remember, to use, intuitive and a simple English word, so I think the best solution is to call it `bag`, which is a common English word and in this case self-explains the use.

P.S.: I noticed that the Node.js 12 CI was not able to test, this is because optional chaining used here is supported only on Node.js 14 and above, so I fixed by checking options object existence and force an empty object if none passed.

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,13 @@
+	 * The extracted number can't be the same as the previous.\
+	 * If the bag is empty, it is reset back to the original numbers.
+	 * @default false
+	 */
```
Please follow the existing doc comment style.

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,13 @@
+	/**
+	 * Puts all numbers in a bag and extracts a random one.\
+	 * The extracted number can't be the same as the previous.\
+	 * If the bag is empty, it is reset back to the original numbers.
```
It should describe some use-cases for this.

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,13 @@
+interface Options {
```
Must be exported.

sindresorhus commented on index.js:
```diff
@@ -1,4 +1,38 @@
-export default function uniqueRandom(minimum, maximum) {
+function filledArray(minimum, maximum) {
```
```suggestion
function integersInRange(minimum, maximum) {
```

sindresorhus commented on index.js:
```diff
@@ -1,4 +1,38 @@
+	return array;
+}
+
+export default function uniqueRandom(minimum, maximum, options) {
+	if (options === undefined) {
+		options = {};
+	}
```
```suggestion
export default function uniqueRandom(minimum, maximum, {bag} = {}) {
```

sindresorhus commented on index.js:
```diff
@@ -1,4 +1,38 @@
+	}
+
+	if (options.bag) {
+		const defaultArray = filledArray(minimum, maximum);
+		let toExtract = [...defaultArray];
+		let previousValue;
+
+		return function random() {
+			if (toExtract.length === 0) {
+				toExtract = [...defaultArray];
+			}
+
+			const value = toExtract[Math.floor(Math.random() * toExtract.length)];
+			if (previousValue === value) {
+				return random();
+			}
+
+			toExtract = toExtract.filter(item => item !== value);
+			previousValue = value;
+			return value;
+		};
```
This should be extracted into an internal top-level function.

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,13 @@
+interface Options {
+	/**
+	 * Puts all numbers in a bag and extracts a random one.\
+	 * The extracted number can't be the same as the previous.\
```
This description does not make it clear how it differs from the existing behavior.

sindresorhus commented:
`bag` is the same as `sack`. A technical term that is meaningless with more context.

sindresorhus commented:
Let's go with `exhaustive` for now and instead include great explanatory docs.

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,22 @@
+export interface Options {
+	/**
+	Prevent extracted numbers regeneration before all possibilities are exhausted.
+
```
> instead include great explanatory docs.

The example helps, but the description needs to be better.

It should in more simple term explain exactly how it works and some common use-cases for it.

Bellisario replied:
Just added a better explanation (further the definition) and an use case (cross-fade image carousel).

sindresorhus commented on index.d.ts:
```diff
@@ -1,3 +1,22 @@
+
+	const random = uniqueRandom(1, 3, {exhaustive: true});
+
+	console.log(random(), random(), random(), "<exhausted>", random(), random(), random());
```
Having a string in the middle there is just confusing and does not help explain how it works.

sindresorhus commented on index.d.ts:
````diff
@@ -1,3 +1,22 @@
+	//=> 1 3 2 <exhausted> 3 1 2
+	```
+	*/
+	readonly exhaustive?: boolean;
````
```js
import uniqueRandom from 'unique-random';

const random = uniqueRandom(1, 3, {exhaustive: true});

console.log(random(), random(), random(), random(), random(), random());
//=> 3
//=> 1
//=> 2
//=> 1 // We exhausted all the numbers in the 1..3 range, so start again and pick from the range 1..3.
//=> 3
//=> 2
```

sindresorhus commented on readme.md:
```diff
@@ -23,10 +23,30 @@ console.log(random(), random(), random());
+
+Type: `boolean`\
+Default: `false`\
+Prevent extracted numbers regeneration before all possibilities are exhausted.
```
This is not in sync with the TS docs.

Bellisario replied:
Just added the example and updates made to TS docs, too.

sindresorhus commented:
There's one more constraint we need to enforce with this option, we need to make sure that the last number in the first round is not the same as the first number in the second round.

We don't want repeat numbers on the boundaries, for example:

```js
import uniqueRandom from 'unique-random';

const random = uniqueRandom(1, 3, {exhaustive: true});

console.log(random(), random(), random(), random(), random(), random());
//=> 3
//=> 1
//=> 2
//=> 2
//=> 3
//=> 1
```

The 2 and 2.

This also needs a test.

Richienb commented on index.js:
```diff
@@ -1,4 +1,39 @@
+
+	return function random() {
+		if (toExtract.length === 0) {
+			toExtract = [...defaultArray];
```
Either:

- Have `integersInRange` return an iterable so there are never 2 arrays stored in memory
- or move values from one array to the other, swapping references when empty

Bellisario replied:
Just updated with reference swap:
every time there a number is extracted, it's stored in `restorer` and deleted from `toExtract`, when empty, swapping `toExtract` with `restorer`.

Bellisario commented:
> There's one more constraint we need to enforce with this option, we need to make sure that the last number in the first round is not the same as the first number in the second round.
> 
> We don't want repeat numbers on the boundaries, for example:
> 
> ```js
> import uniqueRandom from 'unique-random';
> 
> const random = uniqueRandom(1, 3, {exhaustive: true});
> 
> console.log(random(), random(), random(), random(), random(), random());
> //=> 3
> //=> 1
> //=> 2
> //=> 2
> //=> 3
> //=> 1
> ```
> 
> The 2 and 2.
> 
> This also needs a test.

This should already be enforced with:

https://github.com/Bellisario/unique-random/blob/0c9ae5d1a0d45ee7de282aa8a91de87fe4f98333/index.js#L21-L23

and tested with:

https://github.com/Bellisario/unique-random/blob/0c9ae5d1a0d45ee7de282aa8a91de87fe4f98333/test.js#L44

Is there any issue?\
Maybe I missed something and tests can't identify...

Richienb commented on index.js:
```diff
@@ -1,4 +1,41 @@
+			return random();
+		}
+
+		toExtract = toExtract.filter(item => item !== value);
```
I'd rather have the array spliced at the index instead of being filtered for perf

Bellisario replied:
Just updated

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
+		return array;
+	}
+
+	let restorer = [];
+	let toExtract = integersInRange(minimum, maximum);
```
Maybe call these arrays `valuesToPick` and `pickedValues` for consistency

Richienb commented on readme.md:
```diff
@@ -23,10 +23,51 @@ console.log(random(), random(), random());
+##### exhaustive
+
+Type: `boolean`\
+Default: `false`\
```
There should be a newline between the argument type information and default, and the description.

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
-export default function uniqueRandom(minimum, maximum) {
+function _exhaustive(minimum, maximum) {
+	function integersInRange(minimum, maximum) {
```
This function could be moved outside of the function. I'm not sure whether it would be cleaner as a generator.

Bellisario replied:
Done.
I don't think generator may be a good solution, because we're using integersInRange once.

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
+	let previousValue;
+
+	return function random() {
+		if (toExtract.length === 0) {
```
Maybe run this check after the value is drawn so that it isn't run on the first run in which it would never trigger.

Bellisario replied:
Yes, I was thinking this too, but then will be a mess with `picker`, which should reassigned, since we're going to generate it first, then check if "refill" needed, refill and re-extract value...

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
+	return array;
+}
+
+function _exhaustive(minimum, maximum) {
```
Maybe call this function `uniqueRandomExhaustive`

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
+			return random();
+		}
+
+		previousValue = valuesToPick.splice(picker, 1)[0];
```
```suggestion
		[previousValue] = valuesToPick.splice(picker, 1);
```

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
-export default function uniqueRandom(minimum, maximum) {
+function integersInRange(minimum, maximum) {
+	const array = [];
+	for (let index = minimum; index < maximum + 1; index++) {
```
```suggestion
	for (let index = minimum; index <= maximum; index++) {
```

Richienb commented on index.js:
```diff
@@ -1,4 +1,40 @@
+			[valuesToPick, pickedValues] = [pickedValues, valuesToPick];
+		}
+
+		const picker = Math.floor(Math.random() * valuesToPick.length);
```
```suggestion
		const index = Math.floor(Math.random() * valuesToPick.length);
```

Richienb replied:
This will require further updating a bit down

Richienb commented on readme.md:
```diff
@@ -23,10 +23,54 @@ console.log(random(), random(), random());
+Type: `boolean`\
+Default: `false`
+
+Prevent extracted numbers regeneration before all possibilities are exhausted.\
```
```suggestion
Prevent extracted numbers regeneration before all possibilities are exhausted.
```

Richienb replied:
Needs better wording

Bellisario replied:
Maybe this could be more readable:

```suggestion
Prevent extracted numbers from being re-extracted before all possibilities exhausted, ensuring also the last extracted number can't be the first re-extracted.
```

Richienb replied:
We can do better by leaving out some internal implementation details.

Bellisario replied:
Are you intending for example “ensuring also the last extracted number can't be the first re-extracted” for some internal implementation details?

I think instead the definition (first phrase) should be kept so to make sure the user understands well what this option is about, oversimplify may cause the user to misunderstand it.

sindresorhus replied:
I agree. It needs a better wording.

sindresorhus commented:
Bump
